cfme/scripting/requirement.py

```diff
@@ -0,0 +1,336 @@
+        f.writelines(lines)
+
+
+def remove_package_from_file(file_name, package, is_constraint_file=False):
+    """ Removes a package from a requirements file. If the package is not present, does nothing """
+
+    # get the lines
+    with open(file_name, "r") as f:
+        lines = f.readlines()
+
+    if is_constraint_file:
+        # get just the package names
+        mod_lines = [
+            f"{re.split('[=><]', package_name)[0]}\n" for package_name in lines
+        ]
+        index_to_remove = [
+            i for i, package_name in enumerate(mod_lines)
+            if package_name.replace("\n", "") == package
+        ]
+    else:
+        index_to_remove = [
+            i for i, package_name in enumerate(lines) if package_name.replace("\n", "") == package
+        ]
+
+    if len(index_to_remove) > 1:
+        raise ValueError(f"Got more than one package '{package}' to remove from {file_name}. Please"
+                         f"check package name!")
+    if index_to_remove:
+        del lines[index_to_remove[0]]
+        # now write modified lines back to file
+        with open(file_name, "w") as f:
+            f.writelines(lines)
+        click.echo(f"Package '{package}' removed from {file_name}")
+    else:
+        # if index_to_remove is empty, package not found in file
+        click.echo(f"Package '{package}' not found in {file_name}")
+        return
+
+
+# ----- Command line function defs -------
+@click.group(help="Functions for adding, updating, and freezing requirements")
+def main():
+    pass
+
+
+scan_template_opt = click.option(
+    '--scan-template',
+    'scan_template',
+    show_default=True,
+    default=get_rel_path(str(DEFAULT_SCAN_TEMPLATE)),
+    help='The path to the template file (pip -r arg) for scanned imports, will be overwritten',
+)
+
+
+extra_template_opt = click.option(
+    '--extra-template',
+    'extra_template',
+    show_default=True,
+    default=get_rel_path(str(DEFAULT_EXTRA_TEMPLATE)),
+    help='The path to the template file (pip -r arg) for extra packages (e.g. pre-commit)'
+         ', will be overwritten',
+)
+
+constraint_opt = click.option(
+    '--constraint-file',
+    'constraint_file',
+    show_default=True,
+    default=get_rel_path(str(DEFAULT_CONSTRAINT_FILE)),
+    help='The path to the constraint file (pip -r arg) for extra packages (e.g. pre-commit)'
+         ', will be overwritten',
+)
+
+
+frozen_opt = click.option(
+    '--frozen-file',
+    'frozen_file',
+    show_default=True,
+    default=get_rel_path(str(DEFAULT_FROZEN_OUTPUT)),
+    help='The path to the frozen file for ALL imports',
+)
+
+
+package_map_opt = click.option(
+    '--package-map',
+    'package_map',
+    show_default=True,
+    default=get_rel_path(str(DEFAULT_PACKAGE_MAP_FILE)),
+    help='The path to the package map for tricky imports'
+)
+
+
+@main.command(help="Scan repository files for imports from pip-installable packages")
+@scan_template_opt
```

Comment:
What do these decorators do ending in `_opt`? just curious

Reply:
This is only because the `click.option` options we've defined have them. Since at lot of these methods use the same options, we've defined the common options at the top of the file and added them as decorators.

Reply:
Got it.